### PR TITLE
fix: bussin from inside a loop or switch reaches its own function (#319)

### LINKED ISSUE #319: [Bug]: `bussin` inside a loop is caught by the loop's jump buffer — the program dies with "No scope to exit" and prints nothing
### What happens

Returning a value from inside a `flex` or `goon` loop kills the whole program. No output is produced at all, and the exit code is 1.

### Repro

```c
rizz first_match(rizz n) {
    flex (rizz i = 0; i < n; i = i + 1) {
        edgy (i == 2) { bussin i; }
    }
    bussin 0 - 1;
}

skibidi main {
    rizz a = first_match(5);
    yapping("a %d", a);
    yapping("done");
    bussin 0;
}
```

```
$ ./brainrot repro.brainrot
Error: No scope to exit at line 12
$ echo $?
1
```

**stdout is completely empty** — neither `yapping` runs. `goon` behaves identically. Reproduces on `v0.2.0` and on `main` (`f16c9e2`), so it is not a regression from the recent feature work.

### Mechanism

`interpreter_visit_for_statement()` (`interpreter.c:747`) pushes its **own** jump buffer so that `break` can escape the loop:

```c
    PUSH_JUMP_BUFFER();
    if (setjmp(CURRENT_JUMP_BUFFER()) == 0)
    {
        enter_scope();
        ...
        exit_scope();
    }
    POP_JUMP_BUFFER();
```

`handle_return_statement()` (`ast.c`) unwinds to the function scope and then longjmps:

```c
    while (current_scope && !current_scope->is_function_scope) { exit_scope(); }
    if (jump_buffer) { exit_scope(); LONGJMP(); }
```

`LONGJMP()` goes to the **innermost** buffer, which is the loop's, not the function's. So:

1. The return unwinds the loop's scopes and the function scope.
2. The longjmp is caught by the loop, whose `setjmp` now returns non-zero — the `if` body is skipped, `POP_JUMP_BUFFER()` runs, and control falls through to the statement *after* the loop, still inside a function that was supposed to have returned.
3. That function runs on with no scope, and the next `exit_scope()` finds `current_scope == NULL`.
4. `exit_scope()` calls `exit(1)` directly (`ast.c:6949`), so the process dies before anything is flushed.

In short: **`bussin` inside a loop is silently executed as `break`**, and then the mismatched scope stack takes the process down.

### Why it matters

Search-and-return is one of the most ordinary shapes there is:

```c
rizz free_slot(gang Ent *pool, rizz n) {
    flex (rizz i = 0; i < n; i = i + 1) {
        edgy (!pool[i].alive) { bussin i; }
    }
    bussin 0 - 1;
}
```

That is unwritable today. [tung-tung-sahur](https://github.com/Brainrotlang/tung-tung-sahur) works around it by scanning the *whole* pool and keeping the first index in a variable — which reads like a style choice and is not one:

```c
    rizz slot = 0 - 1;
    flex (rizz i = 0; i < n; i = i + 1) {
        edgy (!pool[i].alive) {
            edgy (slot < 0) { slot = i; }
        }
    }
    bussin slot;
```

The failure mode is also about as hostile as it gets: no output, no line number pointing at the actual return, and an error naming an internal concept (`scope`) at the closing brace of `main`.

### Expected

`a 2` then `done`, exit 0.

### Suggested direction

The return needs to reach the *function's* buffer, not the nearest loop's. Options, roughly in increasing order of intrusiveness:

- Have loops record the scope depth at entry and have `handle_return_statement` unwind/rethrow past loop buffers until it reaches a function-scope buffer.
- Give `PUSH_JUMP_BUFFER()` a kind (loop vs function) and have `LONGJMP()` for a return skip loop buffers.
- Separately and regardless: `exit_scope()` calling `exit(1)` means any scope-accounting bug becomes a silent death with no diagnostics flushed. Reporting and unwinding would make the next one of these debuggable.

Related but distinct: #283 (a *bare* `bussin;` is a parse error) — that is the void case and a different failure.

### PR DESCRIPTION
**Pull request by leo-aa88**
## Description

`bussin` inside a `flex`/`goon`/`mewing`/`ohio` killed the process — **empty
stdout**, exit 1, and *"Error: No scope to exit"* pointing at a line that was
not the return.

```c
rizz first_match(rizz n) {
    flex (rizz i = 0; i < n; i = i + 1) {
        edgy (i == 2) { bussin i; }
    }
    bussin 0 - 1;
}
```

### Mechanism

`LONGJMP()` jumps to the **innermost** jump buffer. Seven of the eight push
sites are **break** targets — `for`/`while`/`do-while`/`switch`, duplicated
across the legacy `ast.c` path and the `interpreter.c` visitor path — and
exactly one, `execute_function_call()`, is the **return** target.

So a return from inside a loop landed in the *loop's* `setjmp`, which skipped
its body, popped, and fell through to the statement **after** the loop, inside
a function that was supposed to have returned. `bussin` silently behaved as
`break`. The mismatched scope stack then reached `exit_scope()` with a NULL
scope, which `exit(1)`s directly — so nothing was flushed and the program
printed nothing at all.

### Fix

`JumpBuffer` now records whether it is a function frame, and
`handle_return_statement()` discards intervening break targets before jumping.

Popping there rather than teaching each of the seven loop sites to re-throw
keeps the knowledge in one place — and it is **required, not merely tidier**:
the longjmp abandons those `setjmp` frames, so their own `POP_JUMP_BUFFER()`
never runs and their buffers would leak. The scopes those frames entered are
already gone; the existing loop just above the fix unwinds every non-function
scope first.

## Related Issue

Fixes #319

## Type of Change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Performance improvement
- [ ] Refactor

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have documented my changes in the code or documentation
- [x] I have added tests that prove my change works, at the lowest appropriate layer
- [x] Those tests cover the happy path, the original bug, error cases, edge cases, **and** adversarial cases
- [x] If this PR cannot affect program behavior (docs/comments/license only), I explained that in the Description instead of skipping the items above in silence
- [x] I have run `make format-check` locally (or `make format` to fix)
- [x] I have run the unit tests locally
- [x] I have run the valgrind memory tests locally
- [x] All new and existing tests pass

**Test coverage.** `test_cases/return_from_inside_loop.brainrot` covers `flex`,
`goon`, **two nested loops** (the return has to discard *both* buffers), and
`ohio` in its matching and default arms.

The adversarial part is *what* it asserts. It pins the returned **values**, not
just survival — and that distinction is the whole point: `bussin`-as-`break`
fell through to each function's trailing `bussin 0 - 1;`, so once the scope
stack stopped dying, every one of these would quietly return `-1`. A fixture
checking only for a clean exit would have missed the actual bug entirely.

`bruh` is exercised alongside, since the fix works by *discarding break
targets* — `break` is the thing most at risk of regressing. **Mutation-verified:**
removing the pop fails the fixture.

**Verification.** 482 tests pass (481 → +1). Full valgrind sweep **421/421
clean**. `make format-check` and `make tidy` clean.

## Notes

- **`grind` (continue) is not covered** — it is a parse error (#274), so that
  case could not be written.
- This is the second of #312, #313 and #319. #313 is PR #324; #312 is separate,
  as it touches semantic scoping rather than control flow.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: ast.c [hunk 1]
```diff
@@ -7579,6 +7579,28 @@ void handle_return_statement(ASTNode *expr)
...
+       before anything was flushed, so a search-and-return loop produced
+       no output at all.
+
+       Popping here rather than teaching each loop to re-throw keeps the
+       knowledge in one place: the loops' setjmp frames are abandoned by
+       the longjmp below and their own POP_JUMP_BUFFER() never runs, which
+       is exactly why their buffers have to be released now instead. The
+       scopes those frames entered are already gone -- the loop above
+       unwound every non-function scope before this point. */
+    while (jump_buffer && !jump_buffer->is_function)
```
Comment (leo-aa88) [inline ast.c]: **MAJOR.** This loop assumes a function frame exists somewhere below. In `main` it doesn't — `execute_function_call()` is the only site that pushes one, and `main` isn't called through it. So a `bussin` inside a loop in `main` pops *every* buffer, `jump_buffer` ends up NULL, no `LONGJMP()` happens, and execution falls through into the loop body with its scopes already unwound.

```c
skibidi main {
    yapping("before");
    flex (rizz i = 0; i < 5; i = i + 1) {
        edgy (i == 2) { bussin 0; }
    }
    yapping("AFTER LOOP");
    bussin 0;
}
```

`main`:
```
before
Error: No scope to exit at line 8
exit=1
```

this branch:
```
before
Error: Undefined variable at line 8
Error: Undefined variable in get_expression_type at line 8
Error: Assignment to undefined variable at line 8
Error: No scope to exit at line 8
exit=1
```

Both are wrong — `bussin 0` in `main` should just end the program — but this trades one error for a cascade, because the loop keeps running against a torn-down scope before hitting the same terminal failure. That's the "kept going after the failure and let later passes see the wreckage" shape rather than a new crash, so it's not a blocker; it is a strictly worse diagnostic than what's on `main` today.

The guard is small — only drain when there's something to drain to:

```c
bool has_function_frame = false;
for (JumpBuffer *jb = jump_buffer; jb; jb = jb->next)
    if (jb->is_function) { has_function_frame = true; break; }
if (has_function_frame)
    while (jump_buffer && !jump_buffer->is_function)
        POP_JUMP_BUFFER();
```

That restores `main`'s previous single-error behaviour exactly, and leaves the fix untouched for every function. If you'd rather make `bussin`-in-`main` actually work, that's a bigger question (a plain `bussin 3;` in `main` already exits 0 rather than 3, so the return value isn't plumbed either) and belongs in its own issue — but then it's worth saying so in the comment, since "the loops' setjmp frames are abandoned by the longjmp below" silently assumes the longjmp happens.

Either way the fixture should carry the `main` case; right now every scenario in it is inside a helper function, which is why this didn't show up.

**Review by leo-aa88 — COMMENTED** [review]
# COMMENT

The fix is real and the diagnosis in the commit message is accurate. One shape
it doesn't cover gets *noisier* than before.

**Verified.** Your own fixture on `main`:

```
Error: No scope to exit at line 92
(no other output at all)
```

and on this branch:

```
for      2
while    30
nested   102
switch   11
switch d 99
break    1
done
```

Zero output to six correct lines — the "killed before anything was flushed"
description is exactly right. `bruh` still breaks (the fixture's `break 1`
covers it), and a function called from inside a loop still finds its own frame
rather than the caller's, which is the case the `is_function` flag exists for.

Splitting `PUSH_FUNCTION_JUMP_BUFFER()` out rather than adding a parameter to
the existing macro is the right call — there is exactly one function push site,
and a flag argument at ~15 call sites would be noise.

## The finding

Inline. `bussin` inside a loop in `main` has no function frame to stop at, so
the new loop drains the whole stack and execution falls through with dead
scopes. One error becomes four.

# VERDICT

Good fix on a genuinely nasty bug — `bussin` silently behaving as `bruh` and
then killing the process before flushing is about as bad as a control-flow bug
gets. The remaining case is the one where the invariant the loop assumes ("a
function frame exists below me") doesn't hold, and it's a one-line guard.

File: ast.c (hunk 1 above)
Comment (leo-aa88) [inline ast.c]: Still unguarded at this head, and the rebase onto #320 gave it a second entry point.

`main` is the one function whose body isn't run through `execute_function_call()`, so nothing below this loop is ever `is_function` — it drains the whole stack, `jump_buffer` ends NULL, no `LONGJMP()` fires, and the loop body keeps executing against scopes that were already unwound.

Both spellings now:

```
bussin 0;   inside a loop in main → before / Undefined variable ×3 / No scope to exit
bussin;     inside a loop in main → before / Undefined variable ×3 / No scope to exit
```

`main` branch gives one error for the first and rejects the second at parse time. So the net effect of this PR plus #320 on that shape is: one diagnostic became four, and the bare form joined it.

Same suggestion as before — only drain when there's something to drain to:

```c
bool has_function_frame = false;
for (JumpBuffer *jb = jump_buffer; jb; jb = jb->next)
    if (jb->is_function) { has_function_frame = true; break; }
if (has_function_frame)
    while (jump_buffer && !jump_buffer->is_function)
        POP_JUMP_BUFFER();
```

Strictly non-regressive: every function keeps the fix, `main` keeps its old single-error behaviour. If you'd rather make `bussin` in `main` actually terminate the program, that's the better outcome but a bigger change — a plain `bussin 3;` in `main` already exits 0 rather than 3, so the value isn't plumbed either — and it deserves its own issue rather than being folded in here.

Either way the comment above is worth a sentence: *"the loops' setjmp frames are abandoned by the longjmp below"* assumes the longjmp happens, and in `main` it doesn't. That assumption being implicit is what made this easy to miss.

**Review by leo-aa88 — COMMENTED** [review]
# REQUEST CHANGES

The new commit is good and orthogonal — it doesn't touch the finding, and the
finding got slightly wider.

**What landed.** Composing with #320 is worth having, and the fixture now
covers a bare `bussin;` from inside a loop in a helper function:

```
visiting 0
visiting 1
stopping at 2
done
```

Noticing that #320 landed underneath and that neither PR's fixture covered the
intersection is exactly the right instinct — it's the same "each feature tested
alone, their intersection tested by neither" gap that #315 hit.

**What didn't.** The pop loop at `ast.c:7599` is unchanged, and `main` still
has no function frame to stop at. Both forms now fail the same way — the new
`bussin;` shape inherits it:

```
skibidi main { flex (...) { edgy (i == 2) { bussin 0; } } ... }
→ before / Undefined variable ×3 / No scope to exit

skibidi main { flex (...) { edgy (i == 2) { bussin; } } ... }
→ before / Undefined variable ×3 / No scope to exit
```

versus one error on `main`. So #320's shape widened the surface of the
unguarded case rather than the fix covering it.

483 pass, `format-check` clean.

# VERDICT

Unchanged from my last review: correct fix for functions, and one shape where
the invariant it assumes — that a function frame exists below — doesn't hold.
The guard is a few lines and the `main` case belongs in the fixture, which
still has every scenario inside a helper.

**Review by leo-aa88 — COMMENTED** [review]
# APPROVE

The guard is in and I verified both halves of "strictly non-regressive."

**`main` is back to one error.** Was four:

```
bussin 0;  in a loop in main → before / No scope to exit          (was 4 errors)
bussin;    in a loop in main → before / No scope to exit
```

Identical to what `main` branch produces, which is the bar I asked for.

**The fix is untouched for real functions** — the full fixture still runs
end to end:

```
for 2 / while 30 / nested 102 / switch 11 / switch d 99 / break 1
visiting 0 / visiting 1 / stopping at 2 / done
```

484 collected, `format-check` clean.

The comment addition is the part worth keeping. It names the assumption that
was implicit before — *"abandoned by the longjmp below" assumes the longjmp
happens, and in `main` it does not* — and then says why `main` isn't run
through `execute_function_call()`, with a pointer to `find_symbol()`'s comment
for the grammar reason. That's the second time this session that
`skibidi_function` reducing to a bare statement list has been the root of a
bug; having both comments cross-reference it is how the next person finds it in
one hop instead of three.

Scoping `bussin`-in-`main` out explicitly, with the observation that a plain
`bussin 3;` there already exits 0 rather than 3, is the right call — that's a
value-plumbing question, not a jump-buffer one.

# VERDICT

Correct fix, correct scope, and the comment now carries the assumption that
made the gap invisible. Ship it.

**Review by leo-aa88 — COMMENTED** [review]
# APPROVE (unchanged)

Rebased onto merged `main` (which now carries #327). Pure rebase — verified,
not assumed: every file is byte-identical to the revision I approved.

I checked the `expected_results.json` merge specifically, since #327 added four
entries at the same tail position all three of these PRs append to:

```
valid JSON, 413 entries
lost from main:   []
changed vs main:  []
added:            ['return_from_inside_loop', 'return_from_inside_loop_in_main_fail']
```

Nothing dropped, nothing altered. Worth checking rather than trusting: silently
losing an entry during resolution makes some unrelated fixture stop running
with nothing going red.

Re-confirmed after the rebase: `main`'s `bussin`-in-a-loop still produces the
single `No scope to exit`, and the function fixture still runs to `done`. 488
collected, `format-check` clean.

One correction to my own last review: I asked for the `main` case to be carried
in a fixture, and `return_from_inside_loop_in_main_fail` was already there —
pinning `before` / `No scope to exit`. I'd verified that behaviour by hand and
missed the fixture because I only diffed `ast.c`, `ast.h` and
`return_from_inside_loop.brainrot`. The ask was already satisfied; disregard
that part.

Approval stands.
